py/src/call.py

```diff
@@ -606,22 +663,68 @@ async def do_call(
     from starkware.starkware_utils.commitment_tree.patricia_tree.patricia_tree import (
         PatriciaTree,
     )
+    from starkware.starkware_utils.commitment_tree.patricia_tree.nodes import (
+        EmptyNodeFact,
+    )
     from starkware.starknet.services.api.contract_class import EntryPointType
     from starkware.cairo.lang.vm.crypto import pedersen_hash_func
     from starkware.starknet.testing.state import create_invoke_function
+    from starkware.starknet.storage.starknet_storage import StorageLeaf
+    from starkware.starknet.business_logic.state.objects import (
+        ContractState,
+        ContractCarriedState,
+    )
 
     # hook up the sqlite adapter
     ffc = FactFetchingContext(storage=adapter, hash_func=pedersen_hash_func)
 
     # the root tree has to always be height=251
-    shared_state = SharedState(PatriciaTree(root=root, height=251), block_info)
+    shared_state = SharedState(PatriciaTree(root, 251), block_info)
+
+    # we firstly want to load the target contract
+    contract_addresses = {contract_address}
+    # union all those contracts which have pending updates for
+    contract_addresses = contract_addresses | pending_updates.keys()
+    # exclude pending deploys, which should cause an issue reading them
+    contract_addresses = contract_addresses ^ pending_deployed.keys()
+
+    class_hashes = set()
+
+    # include called contract's class hash to be loaded, it doesn't seem to be always
+    # automatically loaded, see test `test_call_on_pending_deployed`.
+    if contract_address in pending_deployed:
+        class_hashes.add(pending_deployed[contract_address])
+
+    # FIXME: it's unknown if this is really required, but we might end up loading up a lot of classes for nothing
+    for class_hash in pending_deployed.values():
+        class_hashes.add(class_hash)
+
     state_selector = StateSelector(
-        contract_addresses={contract_address}, class_hashes=set()
+        contract_addresses,
+        # we keep setting this to empty because we assume that the pending downloader
+        # has already stored all declared classes locally
+        class_hashes,
```

Comment:
I don't understand this comment? What is empty here? Didn't we just set the `class_hashes`?

Reply:
Outdated, forgotten, will remove

Reply:
Removing soon.